src/JsonSchema/RefResolver.php

```diff
@@ -9,269 +9,175 @@
-            $this->resolveArrayOfSchemas($schema, $propertyName, $sourceUri);
-        }
-
-        // These are all objects containing properties whose values are schemas
-        foreach (array('dependencies', 'patternProperties', 'properties') as $propertyName) {
-            $this->resolveObjectOfSchemas($schema, $propertyName, $sourceUri);
-        }
-
-        --self::$depth;
+        return $this->getRefSchema($jsonPointer, $schema);
     }
 
     /**
-     * Given an object and a property name, that property should be an
-     * array whose values can be schemas.
+     * Recursive resolve schema by traversing through al nodes
      *
-     * @param object $schema       JSON Schema to flesh out
-     * @param string $propertyName Property to work on
-     * @param string $sourceUri    URI where this schema was located
+     * @param object $unresolvedSchema
+     * @param string $fileName
+     * @param array $paths
      */
-    public function resolveArrayOfSchemas($schema, $propertyName, $sourceUri)
+    private function resolveSchemas($unresolvedSchema, $fileName, array $paths)
     {
-        if (! isset($schema->$propertyName) || ! is_array($schema->$propertyName)) {
-            return;
-        }
-
-        foreach ($schema->$propertyName as $possiblySchema) {
-            $this->resolve($possiblySchema, $sourceUri);
+        $objectIterator = new ObjectIterator($unresolvedSchema);
+        foreach ($objectIterator as $toResolveSchema) {
+            if (property_exists($toResolveSchema, '$ref') && is_string($toResolveSchema->{'$ref'})) {
+                $jsonPointer = new JsonPointer($this->uriResolver->resolve($toResolveSchema->{'$ref'}, $fileName));
+                $refSchema = $this->resolveCached((string) $jsonPointer, $paths);
+                $this->unionSchemas($refSchema, $toResolveSchema, $fileName, $paths);
+            }
         }
     }
 
     /**
-     * Given an object and a property name, that property should be an
-     * object whose properties are schema objects.
-     *
-     * @param object $schema       JSON Schema to flesh out
-     * @param string $propertyName Property to work on
-     * @param string $sourceUri    URI where this schema was located
+     * @param JsonPointer $jsonPointer
+     * @param object $schema
+     * @return object
      */
-    public function resolveObjectOfSchemas($schema, $propertyName, $sourceUri)
+    private function getRefSchema(JsonPointer $jsonPointer, $schema)
     {
-        if (! isset($schema->$propertyName) || ! is_object($schema->$propertyName)) {
-            return;
+        /** @var object|array $refSchema */
+        $refSchema = $schema;
+        foreach ($jsonPointer->getPropertyPaths() as $path) {
+            if (is_object($refSchema) && property_exists($refSchema, $path)) {
+                $refSchema = $refSchema->{$path};
+            } elseif (is_array($refSchema) && array_key_exists($path, $refSchema)) {
```

Comment:
The JsonSchema\Entity\JsonPointer should implement the logic as describe in: https://tools.ietf.org/html/rfc6901#section-4